src/fe-text/gui-readline.c

```diff
@@ -647,21 +683,47 @@ static void sig_input(void)
 		unichar key;
 		term_gets(buffer, &line_count);
 		key = g_array_index(buffer, unichar, 0);
+		/* Either Ctrl-k or Ctrl-c is pressed */
 		if (key == 11 || key == 3)
 			paste_flush(key == 11);
 		g_array_free(buffer, TRUE);
 	} else {
 		term_gets(paste_buffer, &paste_line_count);
-		if (paste_detect_time > 0 && paste_buffer->len >= 3) {
+
+		/* use the bracketed paste mode to detect when the user pastes
+		 * some text into the entry */
+		if (paste_bracketed_mode) {
+			int i;
+			int len = paste_buffer->len - G_N_ELEMENTS(bp_end);
+			unichar *ptr = (unichar *) paste_buffer->data;
+
+			if (len <= 0) {
+				return;
+			}
+
+			for (i = 0; i <= len; i++, ptr++) {
+				if (ptr[0] == bp_end[0] && !memcmp(ptr, bp_end, sizeof(bp_end))) {
+					paste_bracketed_end(i, i != len);
+					break;
+				}
+			}
+		}
+		else if (paste_detect_time > 0 && paste_buffer->len >= 3) {
 			if (paste_timeout_id != -1)
 				g_source_remove(paste_timeout_id);
 			paste_timeout_id = g_timeout_add(paste_detect_time, paste_timeout, NULL);
-		} else {
+		} else if (!paste_bracketed_mode) {
 			int i;
 
 			for (i = 0; i < paste_buffer->len; i++) {
 				unichar key = g_array_index(paste_buffer, unichar, i);
 				signal_emit("gui key pressed", 1, GINT_TO_POINTER(key));
+
+				if (paste_bracketed_mode) {
```

Comment:
Yes, see the comment right below `/* just enabled by the signal, remove what was processed so far */`

That is, it was false before, but the signal enabled it.